File: bender_ros/ur/src/ur_node.cpp
/**
 * A simple service server for moving UR around
 */

#include <iostream>
#include <string>
#include <rw/rw.hpp>
#include <rwhw/universalrobots/URCallBackInterface.hpp>
#include <rwhw/universalrobots/UniversalRobotsRTLogging.hpp>

#include <ros/ros.h>
#include <ur/Q.h>
#include <ur/State.h>



using namespace std;
USE_ROBWORK_NAMESPACE;
using namespace rw;
rwhw::URCallBackInterface URInterface;
rwhw::UniversalRobotsRTLogging URRTInterface;



/* MAIN */
int main(int argc, char* argv[])
{
	/* initialize ROS */
	ros::init(argc, argv, "ur_node");
	
	/* initialize node */
	ros::NodeHandle n("~");
	
	/* load parameters from the parameter server */
	string hostip;
	if (!n.getParam("hostip", hostip)) {
		ROS_ERROR("Parameter 'hostip' not found.");
		return -1;
	}
	
	int hostport;
	if (!n.getParam("hostport", hostport)) {
		ROS_ERROR("Parameter 'hostport' not found.");
		return -1;
	}
	
	string robotip;
	if (!n.getParam("robotip", robotip)) {
		ROS_ERROR("Parameter 'robotip' not found.");
		return -1;
	}
	
	int robotport;
	if (!n.getParam("robotport", robotport)) {
		ROS_ERROR("Parameter 'robotport' not found.");
		return -1;
	}
	
	int robotrtport;
	if (!n.getParam("robotrtport", robotrtport)) {
		ROS_ERROR("Parameter 'robotport' not found.");
		return -1;
	}
	
	string script;
	if (!n.getParam("script", script)) {
		ROS_ERROR("Parameter 'script' not found.");
		return -1;
	}
	
	int rate;
	if (!n.getParam("rate", rate)) {
		ROS_ERROR("Parameter 'rate' not found.");
		return -1;
	}
	
	/* establish robot communication */
	URRTInterface.connect(robotip, robotrtport);
	URRTInterface.start();
	URInterface.connect(robotip, robotport);
	URInterface.startCommunication(hostip, hostport, script);
	
	/* advertise services */
	/*ros::ServiceServer urGetQService = nh.advertiseService("ur_get_q", ur_get_q);
	ros::ServiceServer urStopService = nh.advertiseService("ur_stop", ur_stop);
	ros::ServiceServer urMoveQService = nh.advertiseService("ur_move_to_q", ur_move_to_q);
	ros::ServiceServer urMoveTService = nh.advertiseService("ur_move_to_t", ur_move_to_t);
	ros::ServiceServer urServoQService = nh.advertiseService("ur_servo_to_q", ur_servo_to_q);*/
	
	/* advertise topics */
	ros::Publisher statePub = n.advertise<ur::State>("state", 1000);

	ROS_INFO("UR node ready.");
	
	/* publish state topic */
	ros::Rate loop_rate(rate);
	while (ros::ok()) {
		/* get last RT data */
		rwhw::URRTData URData;
		URData = URRTInterface.getLastData();
		rw::math::Q qActual = URData.qActual;
		
		/* create message and publish*/
		ur::State msg;
				
		statePub.publish(msg);
		
		/* sleep */
		ros::spinOnce();
		loop_rate.sleep();
	}
	
	return 0;
}

File: bender_ros/ur/src/ur_node.py
#!/usr/bin/python

import universal_robot
import rospy
import std_msgs.msg

from ur.msg import *
from ur.srv import *


class URNode:
  def __init__(self):
    # initialize node
    rospy.init_node('ur_node')

    # load parameters
    self.robot_ip = rospy.get_param('~robot_ip')
    self.robot_port = rospy.get_param('~robot_port')
    self.robot_data_port = rospy.get_param('~robot_data_port')
    self.rate = float(rospy.get_param('~rate'))

    # connect to robot
    self.robot = universal_robot.UR(self.robot_ip, self.robot_port, self.robot_data_port)
    self.robot.connect()

    # start state publisher
    self.topic = '~state'
    self.state_pub = rospy.Publisher(self.topic, State, queue_size=100) # todo: queue_size?
  
    # advertise services
    self.home_srv = rospy.Service('~home', Home, self.home)
    self.moveq_srv = rospy.Service('~move_q', MoveQ, self.move_q)
    self.movet_srv = rospy.Service('~move_t', MoveT, self.move_t)

  def home(self, req):
    """Moves robot to home (upright) position."""
    self.robot.move_q([0, -1.57, 0, 0, 0, 0], 0.1) # todo: hardcoded values
    return HomeResponse()

  def move_q(self, req):
    """Handles the move_q service."""
    target = req.target.data
    speed = req.speed
    self.robot.move_q(target, speed)
    return MoveQResponse()

  def move_t(self, req):
    """Handles the move_t service."""
    target = req.target.data
    speed = req.speed
    self.robot.move_t(target, speed)
    return MoveTResponse()
  
  def run(self):
    while not rospy.is_shutdown():
      # todo: add other information to the state topic
      
      # poll robot
      q = self.robot.get_q()
      pos = self.robot.get_pos()
      
      # create message
      state_msg = State()
      state_msg.header = std_msgs.msg.Header()
      state_msg.header.stamp = rospy.Time.now()
      state_msg.q.data = q
      state_msg.pos.data = pos
      self.state_pub.publish(state_msg)

      rospy.sleep(1/self.rate)


def main():
  node = URNode()
  node.run()


if __name__ == '__main__':
  main()

